syntax = "proto3";

package gradient;

// GradientTensor represents a single gradient tensor
message GradientTensor {
  // Tensor data as flat array
  repeated float data = 1 [packed = true];
  
  // Original shape of the tensor
  repeated int64 shape = 2 [packed = true];
  
  // Data type (e.g., "float32", "float16")
  string dtype = 3;
  
  // Parameter name (e.g., "lora_A.0.weight")
  string name = 4;
}

// GradientPackage contains all gradients and metadata
message GradientPackage {
  // List of gradient tensors
  repeated GradientTensor gradients = 1;
  
  // Metadata dictionary (key-value pairs)
  map<string, string> metadata = 2;
  
  // Training round ID
  uint64 training_round_id = 3;
  
  // Miner address
  string miner_address = 4;
  
  // Gradient hash
  string gradient_hash = 5;
  
  // Timestamp
  int64 timestamp = 6;
}

